File: src/main/java/fi/finlit/edith/sql/domain/Note.java
package fi.finlit.edith.sql.domain;

import java.util.HashSet;
import java.util.Set;
import java.util.regex.Pattern;

import javax.persistence.ElementCollection;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;

import org.apache.commons.lang.StringUtils;

@Entity
public class Note {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    private String lemma;

    @ManyToOne
    private Term term;

    private Long editedOn;

    @ManyToMany(fetch = FetchType.LAZY)
    private Set<User> allEditors = new HashSet<User>();

    @OneToMany(fetch = FetchType.LAZY, mappedBy="note")
    private Set<NoteComment> comments = new HashSet<NoteComment>();

    @ManyToOne
    private User lastEditedBy;

    private String sources;

    private String subtextSources;

    private String description;

    @Enumerated(EnumType.STRING)
    private NoteStatus status = NoteStatus.INITIAL;

    @ElementCollection(fetch=FetchType.LAZY)
    @Enumerated(EnumType.STRING)
    private Set<NoteType> types = new HashSet<NoteType>();

    @Enumerated(EnumType.STRING)
    private NoteFormat format;

    private String lemmaMeaning;

    @ManyToOne
    private Person person;

    @ManyToOne
    private Place place;

//    private int documentNoteCount = 0;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getLemma() {
        return lemma;
    }

    public void setLemma(String lemma) {
        this.lemma = lemma;
    }

    public Term getTerm() {
        return term;
    }

    public void setTerm(Term term) {
        this.term = term;
    }

    public Long getEditedOn() {
        return editedOn;
    }

    public void setEditedOn(Long editedOn) {
        this.editedOn = editedOn;
    }

    public Set<User> getAllEditors() {
        return allEditors;
    }

    public void setAllEditors(Set<User> allEditors) {
        this.allEditors = allEditors;
    }

    public Set<NoteComment> getComments() {
        return comments;
    }

    public void setComments(Set<NoteComment> comments) {
        this.comments = comments;
    }

    public User getLastEditedBy() {
        return lastEditedBy;
    }

    public void setLastEditedBy(User lastEditedBy) {
        this.lastEditedBy = lastEditedBy;
    }

    public String getSources() {
        return sources;
    }

    public void setSources(String sources) {
        this.sources = sources;
    }

    public String getSubtextSources() {
        return subtextSources;
    }

    public void setSubtextSources(String subtextSources) {
        this.subtextSources = subtextSources;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public NoteStatus getStatus() {
        return status;
    }

    public void setStatus(NoteStatus status) {
        this.status = status;
    }

    public Set<NoteType> getTypes() {
        return types;
    }

    public void setTypes(Set<NoteType> types) {
        this.types = types;
    }

    public NoteFormat getFormat() {
        return format;
    }

    public void setFormat(NoteFormat format) {
        this.format = format;
    }

    public String getLemmaMeaning() {
        return lemmaMeaning;
    }

    public void setLemmaMeaning(String lemmaMeaning) {
        this.lemmaMeaning = lemmaMeaning;
    }

//    public int getDocumentNoteCount() {
//        return documentNoteCount;
//    }
//
//    public void setDocumentNoteCount(int documentNoteCount) {
//        this.documentNoteCount = documentNoteCount;
//    }

    private static final Pattern WHITESPACE = Pattern.compile("\\s+");

    public static String createLemmaFromLongText(String text) {
        String result = null;
        if (WHITESPACE.matcher(text).find()) {
            String[] words = StringUtils.split(text);
            if (words.length == 2) {
                result = words[0] + " " + words[1];
            } else if (words.length > 1) {
                result = words[0] + " \u2013 \u2013 " + words[words.length - 1];
            } else {
                result = words[0];
            }
        } else {
            result = text;
        }
        return result;
    }

}

File: src/main/scala/fi/finlit/edith/sql/domain/Note.scala
package fi.finlit.edith.sql.domain

import java.util.ArrayList
import java.util.Collection
import java.util.HashSet
import java.util.Set
import java.util.regex.Pattern
import javax.persistence.Column
import javax.persistence.ElementCollection
import javax.persistence.Entity
import javax.persistence.EnumType
import javax.persistence.Enumerated
import javax.persistence.FetchType
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.JoinTable
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table
import org.apache.commons.lang.StringUtils
import org.hibernate.annotations.Cache
import org.hibernate.annotations.CacheConcurrencyStrategy
import org.hibernate.annotations.Cascade
import org.hibernate.annotations.CascadeType
import org.joda.time.DateTime
import fi.finlit.edith.Identifiable
import Note._
import scala.reflect.BeanProperty
//remove if not needed
import scala.collection.JavaConversions._

object Note {

  private val WHITESPACE = Pattern.compile("\\s+")

  def createLemmaFromLongText(text: String): String = {
    var result: String = null
    if (WHITESPACE.matcher(text).find()) {
      var words = StringUtils.split(text)
      if (words.length == 2) {
        result = words(0) + " " + words(1)
      } else if (words.length > 1) {
        result = words(0) + " – – " + words(words.length - 1)
      } else {
        result = words(0)
      }
    } else {
      result = text
    }
    result
  }
}

@Entity
@Table(name = "note")
@Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
class Note extends Identifiable {

  @Id
  @GeneratedValue(strategy = GenerationType.AUTO)
  @BeanProperty
  var id: java.lang.Long = _

  @BeanProperty
  var lemma: String = _

  @ManyToOne
  @Cascade(value = CascadeType.SAVE_UPDATE)
  @BeanProperty
  var term: Term = _

  @BeanProperty
  var editedOn: java.lang.Long = _

  @ManyToMany(fetch = FetchType.LAZY)
  @BeanProperty
  var allEditors = new HashSet[User]()

  @OneToMany(fetch = FetchType.LAZY, mappedBy = "note")
  @BeanProperty
  var comments = new HashSet[NoteComment]()

  @ManyToOne
  @BeanProperty
  var lastEditedBy: User = _

  @BeanProperty
  var sources: String = _

  @BeanProperty
  var subtextSources: String = _

  @BeanProperty
  var deleted: Boolean = _

  @Column(columnDefinition = "TEXT")
  @BeanProperty
  var description: String = _

  @Enumerated(EnumType.STRING)
  @BeanProperty
  var status = NoteStatus.INITIAL

  @ElementCollection(fetch = FetchType.LAZY)
  @Enumerated(EnumType.STRING)
  @JoinTable(name = "note_types")
  @BeanProperty
  var types = new HashSet[NoteType]()

  @Enumerated(EnumType.STRING)
  @BeanProperty
  var format: NoteFormat = _

  @BeanProperty
  var lemmaMeaning: String = _

  @ManyToOne
  @Cascade(value = CascadeType.SAVE_UPDATE)
  @BeanProperty
  var person: Person = _

  @ManyToOne
  @Cascade(value = CascadeType.SAVE_UPDATE)
  @BeanProperty
  var place: Place = _

  @BeanProperty
  var documentNoteCount = 0

  def getEditedOnDate(): DateTime = new DateTime(editedOn)

  def addComment(comment: NoteComment) {
    comments.add(comment)
  }

  def removeComment(comment: NoteComment) {
    comments.remove(comment)
  }

  def decDocumentNoteCount() {
    if (documentNoteCount > 0) {
      documentNoteCount -= 1
    }
  }

  def incDocumentNoteCount() {
    documentNoteCount += 1
  }

  def addEditor(user: User) {
    allEditors.add(user)
  }

  def getEditors(): String = {
    var result = new ArrayList[String]()
    for (user <- getAllEditors if getLastEditedBy == user) {
      result.add(user.getUsername)
    }
    StringUtils.join(result, ", ")
  }
}
